File: src/app/use_cases/page.py
from dataclasses import dataclass


@dataclass(frozen=True)
class Page[T]:
    """Uma fatia de resultados e o total que casa com a consulta."""

    items: list[T]
    page: int
    per_page: int
    total: int

    @property
    def total_pages(self) -> int:
        if self.per_page <= 0:
            return 0
        return -(-self.total // self.per_page)

File: frontend/react/src/domain/page.ts
export interface Page<T> {
  items: T[];
  page: number;
  perPage: number;
  total: number;
  totalPages: number;
}

export function isLastPage<T>(page: Page<T>): boolean {
  return page.page >= page.totalPages;
}

export function isEmpty<T>(page: Page<T>): boolean {
  return page.total === 0;
}
